File: workflows/org.linxira.benchmark-r/src/benchmark_harness.R
#!/usr/bin/env Rscript
# Benchmark harness: the R backend of `linxira-bio benchmark run`.
#
# The worker invokes this pack when a request carries execution.backend = "r".
# It runs the independent R implementation of the requested capability (see
# implementations/), writes the same V2 result envelope the Rust engine
# produces, and self-reports the in-process wall time and peak RSS
# (/proc/self/status VmHWM) as an `info` diagnostic so the report can disclose
# interpreter start-up separately from the analysis (ROADMAP M2-T3, §7.1).

PACK_ID <- "org.linxira.benchmark-r"
PACK_VERSION <- "0.1.0"
BACKEND <- "r"
SELF_REPORTED_CODE <- "benchmark.self_reported"
HARNESS_PACKAGES <- c(jsonlite = ">=1.8.9,<3.0.0", digest = ">=0.6.37,<0.7.0")

core_version <- function() {
  value <- Sys.getenv("LINXIRA_BIO_CORE_VERSION", unset = "")
  if (nzchar(value)) value else "unknown"
}

utc_now <- function() format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC")

request_error <- function(message) {
  stop(structure(
    list(message = message, call = NULL),
    class = c("request_error", "error", "condition")
  ))
}

implementation_error <- function(message) {
  stop(structure(
    list(message = message, call = NULL),
    class = c("implementation_error", "error", "condition")
  ))
}

require_object <- function(value, context) {
  if (!is.list(value) || (length(value) > 0L && is.null(names(value)))) {
    request_error(sprintf("%s must be an object", context))
  }
  value
}

require_string <- function(value, context) {
  if (!is.character(value) || length(value) != 1L || is.na(value) || !nzchar(value)) {
    request_error(sprintf("%s must be a non-empty string", context))
  }
  value
}

is_array <- function(value) is.list(value) && is.null(names(value))

same_path <- function(left, right) {
  canonical <- function(path) {
    path <- normalizePath(path, winslash = "/", mustWork = FALSE)
    # normalizePath leaves non-existent paths untouched, so unify separators
    # and trailing slashes by hand before comparing.
    sub("/+$", "", chartr("\\", "/", path))
  }
  identical(canonical(left), canonical(right))
}

# Benchmark policy: a project-isolated library is honoured when configured,
# otherwise the interpreter's own library paths are used, because the point of
# this backend is to measure the R stack as installed on the benchmark host.
# Whatever resolves is disclosed in provenance.software.
configure_library <- function() {
  configured <- Sys.getenv("LINXIRA_BIO_WORKFLOW_R_LIBRARY", unset = "")
  if (!nzchar(configured)) {
    return(invisible(NULL))
  }
  if (!dir.exists(configured)) {
    stop(sprintf("LINXIRA_BIO_WORKFLOW_R_LIBRARY names a missing directory: %s", configured))
  }
  library <- normalizePath(configured, winslash = "/", mustWork = TRUE)
  .libPaths(c(library, .libPaths()))
  invisible(library)
}

version_satisfies <- function(actual, requirement) {
  clauses <- trimws(strsplit(requirement, ",", fixed = TRUE)[[1L]])
  actual <- package_version(actual)
  for (clause in clauses) {
    operator <- regmatches(clause, regexpr("^(>=|<=|==|<|>)", clause))
    bound <- package_version(sub("^(>=|<=|==|<|>)", "", clause))
    ok <- switch(operator,
      ">=" = actual >= bound, "<=" = actual <= bound, "==" = actual == bound,
      "<" = actual < bound, ">" = actual > bound, FALSE
    )
    if (!ok) return(FALSE)
  }
  TRUE
}

check_packages <- function(requirements) {
  drift <- list()
  for (package in names(requirements)) {
    if (!requireNamespace(package, quietly = TRUE)) {
      stop(sprintf("dependency %s %s is not installed", package, requirements[[package]]))
    }
    installed <- as.character(utils::packageVersion(package))
    if (!version_satisfies(installed, requirements[[package]])) {
      drift[[length(drift) + 1L]] <- list(
        code = "dependency_version_drift",
        severity = "warning",
        message = sprintf(
          "%s %s is installed; the pack lock requires %s. Numbers remain comparable but the environment disclosure records the installed version.",
          package, installed, requirements[[package]]
        )
      )
    }
  }
  drift
}

sha256_file <- function(path) digest::digest(path, algo = "sha256", file = TRUE)

# Peak resident set size of this process in MiB, or NULL where the kernel
# does not expose it (Windows, macOS): the report then keeps the field null
# rather than inventing a figure.
peak_rss_mb <- function() {
  status <- "/proc/self/status"
  if (!file.exists(status)) return(NULL)
  lines <- readLines(status, warn = FALSE)
  line <- grep("^VmHWM:", lines, value = TRUE)
  if (length(line) != 1L) return(NULL)
  fields <- strsplit(trimws(sub("^VmHWM:", "", line)), "[[:space:]]+")[[1L]]
  kilobytes <- suppressWarnings(as.numeric(fields[[1L]]))
  if (is.na(kilobytes)) return(NULL)
  kilobytes / 1024
}

load_implementations <- function(directory) {
  files <- sort(list.files(directory, pattern = "\\.R$", full.names = TRUE))
  registry <- list()
  for (file in files) {
    environment <- new.env(parent = globalenv())
    sys.source(file, envir = environment)
    implementation <- get("IMPLEMENTATION", envir = environment)
    registry[[implementation$capability]] <- implementation
  }
  registry
}

validate_request <- function(document, result_path, registry) {
  request <- require_object(document, "request")
  if (!identical(request$schema_version, "2")) request_error("request schema_version must be \"2\"")
  job_id <- require_string(request$job_id, "job_id")
  capability <- require_string(request$capability, "capability")
  implementation <- registry[[capability]]
  if (is.null(implementation)) {
    request_error(sprintf(
      "%s has no r implementation of %s; supported: %s",
      PACK_ID, capability, paste(sort(names(registry)), collapse = ", ")
    ))
  }
  execution <- require_object(request$execution, "execution")
  backend <- if (is.null(execution$backend)) BACKEND else execution$backend
  if (!identical(backend, BACKEND)) {
    request_error(sprintf("execution.backend must be \"%s\", got \"%s\"", BACKEND, backend))
  }

  if (!is_array(request$inputs)) request_error("inputs must be an array")
  resolved <- list()
  declared_sha256 <- list()
  for (index in seq_along(request$inputs)) {
    artifact <- require_object(request$inputs[[index]], sprintf("inputs[%d]", index - 1L))
    role <- require_string(artifact$role, sprintf("inputs[%d].role", index - 1L))
    if (!role %in% implementation$input_roles) {
      request_error(sprintf("%s does not accept input role %s", capability, role))
    }
    if (!is.null(resolved[[role]])) request_error(sprintf("duplicate input role: %s", role))
    if (!identical(artifact$cardinality, "single")) {
      request_error(sprintf("input role %s requires cardinality \"single\"", role))
    }
    if (!is_array(artifact$files) || length(artifact$files) != 1L) {
      request_error(sprintf("input role %s requires exactly one file", role))
    }
    file <- require_object(artifact$files[[1L]], sprintf("inputs[%d].files[0]", index - 1L))
    path <- require_string(file$path, sprintf("inputs[%d].files[0].path", index - 1L))
    if (!file.exists(path) || dir.exists(path)) request_error(sprintf("input file does not exist: %s", path))
    compression <- if (is.null(file$compression)) "none" else file$compression
    if (!identical(compression, "none")) {
      request_error("the benchmark harness reads uncompressed inputs only")
    }
    sha <- file$sha256
    if (!is.null(sha) && (!is.character(sha) || length(sha) != 1L || nchar(sha) != 64L)) {
      request_error(sprintf("inputs[%d].files[0].sha256 must be a 64-character hex string", index - 1L))
    }
    resolved[[role]] <- normalizePath(path, winslash = "/", mustWork = TRUE)
    declared_sha256[[role]] <- if (is.null(sha)) NA_character_ else tolower(sha)
  }
  missing <- setdiff(implementation$input_roles, names(resolved))
  if (length(missing) > 0L) {
    request_error(sprintf("%s requires inputs: %s", capability, paste(missing, collapse = ", ")))
  }

  parameters <- request$parameters
  if (is.null(parameters)) parameters <- list()
  parameters <- require_object(parameters, "parameters")
  output_directory <- require_string(parameters$output_directory, "parameters.output_directory")
  if (!grepl("^(/|[A-Za-z]:[/\\\\]|\\\\\\\\)", output_directory)) {
    request_error("parameters.output_directory must be an absolute path")
  }
  for (name in names(parameters)) {
    if (name != "output_directory" && !name %in% implementation$parameters) {
      request_error(sprintf("%s does not accept parameter %s", capability, name))
    }
  }
  if (!same_path(dirname(result_path), output_directory)) {
    request_error("the result path must live directly inside parameters.output_directory")
  }
  if (!dir.exists(dirname(output_directory))) {
    request_error(sprintf("output parent directory does not exist: %s", dirname(output_directory)))
  }
  for (path in resolved) {
    if (same_path(path, output_directory)) request_error("output directory must differ from every input")
  }

  list(
    job_id = job_id,
    capability = capability,
    implementation = implementation,
    inputs = resolved,
    declared_sha256 = declared_sha256,
    parameters = parameters[setdiff(names(parameters), "output_directory")],
    output_directory = output_directory,
    result_path = result_path
  )
}

write_json_file <- function(path, document) {
  jsonlite::write_json(
    document, path,
    auto_unbox = TRUE, null = "null", na = "null", digits = NA, pretty = TRUE
  )
}

run_benchmarked <- function(config, started_at) {
  implementation <- config$implementation
  capability <- config$capability
  output_directory <- config$output_directory
  drift <- check_packages(implementation$packages)

  # The worker already verified every declared hash and re-checks the inputs
  # after the pack exits; the declared digest is reused so a multi-terabyte
  # input is not hashed a third time inside this backend's wall time.
  input_sha256 <- list()
  for (role in names(config$inputs)) {
    declared <- config$declared_sha256[[role]]
    input_sha256[[role]] <- if (is.na(declared)) sha256_file(config$inputs[[role]]) else declared
  }

  # Instrumentation brackets the analysis only; interpreter start-up,
  # validation and output writing are visible to the outer timer instead.
  started <- Sys.time()
  result <- implementation$run(config$inputs, config$parameters)
  wall_ms <- as.numeric(difftime(Sys.time(), started, units = "secs")) * 1000
  peak <- peak_rss_mb()

  software <- c(
    list(list(name = "R", version = paste(R.version$major, R.version$minor, sep = "."))),
    implementation$software(),
    lapply(names(HARNESS_PACKAGES), function(package) {
      list(name = package, version = as.character(utils::packageVersion(package)), package_id = paste0("cran-", package))
    })
  )
  diagnostics <- c(drift, list(list(
    code = SELF_REPORTED_CODE,
    severity = "info",
    message = as.character(jsonlite::toJSON(
      list(
        wall_ms = round(wall_ms, 3),
        peak_rss_mb = if (is.null(peak)) NULL else round(peak, 3),
        instrument = "Sys.time + /proc/self/status VmHWM",
        backend = BACKEND
      ),
      auto_unbox = TRUE, null = "null", digits = NA
    ))
  )))

  staging <- tempfile(pattern = ".linxira-benchmark-", tmpdir = dirname(output_directory))
  if (!dir.create(staging, mode = "0700")) stop("could not create the staging directory")
  committed <- FALSE
  on.exit(if (!committed && dir.exists(staging)) unlink(staging, recursive = TRUE, force = TRUE), add = TRUE)

  artifact_name <- paste0(capability, ".result.json")
  staged_artifact <- file.path(staging, artifact_name)
  write_json_file(staged_artifact, result)
  lock_path <- file.path(SCRIPT_DIRECTORY, "..", "dependencies.lock.json")
  envelope <- list(
    schema_version = "2",
    job_id = config$job_id,
    capability = capability,
    status = "ok",
    result = result,
    artifacts = list(list(
      artifact_id = "benchmark-result",
      role = "benchmark-result",
      kind = "report",
      path = file.path(output_directory, artifact_name),
      format = "json",
      media_type = "application/json",
      size_bytes = as.numeric(file.info(staged_artifact)$size),
      sha256 = sha256_file(staged_artifact)
    )),
    provenance = list(
      engine_version = PACK_VERSION,
      execution_mode = "local-cpu",
      core_version = core_version(),
      started_at = started_at,
      finished_at = utc_now(),
      software = software,
      input_sha256 = input_sha256,
      command = list("Rscript", "src/benchmark_harness.R", "--request", "<request>", "--result", "<result>"),
      dependency_lock_sha256 = sha256_file(normalizePath(lock_path, winslash = "/", mustWork = TRUE))
    ),
    diagnostics = diagnostics
  )
  write_json_file(file.path(staging, basename(config$result_path)), envelope)
  if (dir.exists(output_directory) || file.exists(output_directory)) {
    request_error("output directory appeared while the analysis was running")
  }
  if (!file.rename(staging, output_directory)) stop("could not activate the output directory")
  committed <- TRUE
  envelope
}

minimal_error_json <- function(job_id, capability, message, started_at) {
  sprintf(
    paste0(
      '{"schema_version":"2","job_id":%s,"capability":%s,"status":"error",',
      '"result":{},"artifacts":[],"provenance":{"engine_version":"%s",',
      '"execution_mode":"local-cpu","core_version":"%s",',
      '"started_at":"%s","finished_at":"%s"},',
      '"diagnostics":[{"code":"workflow_failed","severity":"error","message":%s}]}'
    ),
    encodeString(job_id, quote = '"'), encodeString(capability, quote = '"'),
    PACK_VERSION, core_version(), started_at, utc_now(),
    encodeString(message, quote = '"')
  )
}

write_error_json_atomic <- function(result_path, payload) {
  target <- normalizePath(result_path, winslash = "/", mustWork = FALSE)
  parent <- dirname(target)
  if (dir.exists(parent)) {
    if (file.exists(target) || dir.exists(target)) return(FALSE)
    temporary <- tempfile(pattern = ".linxira-error-", tmpdir = parent)
    writeLines(enc2utf8(payload), temporary, useBytes = TRUE)
    if (!file.rename(temporary, target)) {
      unlink(temporary, force = TRUE)
      return(FALSE)
    }
    return(TRUE)
  }
  grandparent <- dirname(parent)
  if (!dir.exists(grandparent) || file.exists(parent)) return(FALSE)
  staging <- tempfile(pattern = ".linxira-error-", tmpdir = grandparent)
  if (!dir.create(staging, mode = "0700")) return(FALSE)
  committed <- FALSE
  on.exit(if (!committed && dir.exists(staging)) unlink(staging, recursive = TRUE, force = TRUE), add = TRUE)
  writeLines(enc2utf8(payload), file.path(staging, basename(target)), useBytes = TRUE)
  if (!file.rename(staging, parent)) return(FALSE)
  committed <- TRUE
  TRUE
}

parse_arguments <- function(arguments) {
  usage <- "usage: benchmark_harness.R --request REQUEST --result RESULT"
  if (length(arguments) != 4L) stop(usage)
  request_index <- which(arguments == "--request")
  result_index <- which(arguments == "--result")
  if (length(request_index) != 1L || length(result_index) != 1L ||
      request_index == length(arguments) || result_index == length(arguments)) {
    stop(usage)
  }
  if (!setequal(c(request_index, result_index, request_index + 1L, result_index + 1L), seq_along(arguments))) {
    stop("unsupported or duplicate command-line arguments")
  }
  list(request = arguments[[request_index + 1L]], result = arguments[[result_index + 1L]])
}

main <- function(arguments = commandArgs(trailingOnly = TRUE)) {
  started_at <- utc_now()
  job_id <- "workflow-error"
  capability <- "unknown"
  result_path <- NULL
  tryCatch({
    options <- parse_arguments(arguments)
    result_path <- options$result
    configure_library()
    if (!file.exists(options$request) || dir.exists(options$request)) {
      request_error(sprintf("request file does not exist: %s", options$request))
    }
    if (same_path(options$request, options$result)) {
      request_error("result path must not alias the request file")
    }
    check_packages(HARNESS_PACKAGES)
    document <- jsonlite::fromJSON(options$request, simplifyVector = FALSE)
    if (is.list(document)) {
      if (is.character(document$job_id) && length(document$job_id) == 1L && nzchar(document$job_id)) {
        job_id <- document$job_id
      }
      if (is.character(document$capability) && length(document$capability) == 1L && nzchar(document$capability)) {
        capability <- document$capability
      }
    }
    registry <- load_implementations(file.path(SCRIPT_DIRECTORY, "implementations"))
    config <- validate_request(document, options$result, registry)
    run_benchmarked(config, started_at)
    0L
  }, error = function(error) {
    payload <- minimal_error_json(job_id, capability, conditionMessage(error), started_at)
    if (!is.null(result_path)) {
      try(write_error_json_atomic(result_path, payload), silent = TRUE)
    }
    message(sprintf("%s: %s", PACK_ID, conditionMessage(error)))
    2L
  })
}

arguments_all <- commandArgs(trailingOnly = FALSE)
file_argument <- grep("^--file=", arguments_all, value = TRUE)
SCRIPT_DIRECTORY <- if (length(file_argument) == 1L) {
  dirname(normalizePath(sub("^--file=", "", file_argument), winslash = "/", mustWork = TRUE))
} else {
  getwd()
}

if (sys.nframe() == 0L) quit(status = main(), save = "no")

File: workflows/org.linxira.benchmark-python/src/benchmark_harness.py
#!/usr/bin/env python3
"""Benchmark harness: the Python backend of `linxira-bio benchmark run`.

The worker invokes this pack when a request carries
``execution.backend = "python"``. It runs the independent Python
implementation of the requested capability (see ``implementations/``), writes
the same V2 result envelope the Rust engine produces, and self-reports the
in-process wall time and peak RSS as an ``info`` diagnostic so the report can
disclose interpreter start-up separately from the analysis itself
(ROADMAP M2-T3, methodology §7.1).
"""

from __future__ import annotations

import argparse
import hashlib
import json
import os
import shutil
import sys
import tempfile
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parent))

from implementations import IMPLEMENTATIONS  # noqa: E402

PACK_ID = "org.linxira.benchmark-python"
PACK_VERSION = "0.1.0"
BACKEND = "python"
SELF_REPORTED_CODE = "benchmark.self_reported"
LOCKED_BIOPYTHON = "1.85"


class RequestError(ValueError):
    """A stable, user-correctable request validation failure."""


def utc_now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds").replace("+00:00", "Z")


def core_version() -> str:
    return os.environ.get("LINXIRA_BIO_CORE_VERSION", "unknown")


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def peak_rss_mb() -> float | None:
    """Peak resident set size of this process, or None where unavailable.

    ``resource`` does not exist on Windows; the report then keeps the field
    null instead of inventing a number (methodology §7.1: no fabricated I/O
    or memory figures).
    """
    try:
        import resource
    except ImportError:
        return None
    usage = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    if sys.platform == "darwin":
        return usage / (1024.0 * 1024.0)
    return usage / 1024.0


def require_object(value: Any, context: str) -> dict[str, Any]:
    if type(value) is not dict:
        raise RequestError(f"{context} must be an object")
    return value


def require_string(value: Any, context: str) -> str:
    if type(value) is not str or not value:
        raise RequestError(f"{context} must be a non-empty string")
    return value


def validate_request(document: Any, result_path: Path) -> dict[str, Any]:
    request = require_object(document, "request")
    if request.get("schema_version") != "2":
        raise RequestError("request schema_version must be \"2\"")
    job_id = require_string(request.get("job_id"), "job_id")
    capability = require_string(request.get("capability"), "capability")
    implementation = IMPLEMENTATIONS.get(capability)
    if implementation is None:
        supported = ", ".join(sorted(IMPLEMENTATIONS))
        raise RequestError(
            f"{PACK_ID} has no python implementation of {capability}; supported: {supported}"
        )
    execution = require_object(request.get("execution"), "execution")
    backend = execution.get("backend", BACKEND)
    if backend != BACKEND:
        raise RequestError(f"execution.backend must be {BACKEND!r}, got {backend!r}")

    inputs = request.get("inputs")
    if type(inputs) is not list:
        raise RequestError("inputs must be an array")
    resolved: dict[str, Path] = {}
    declared_sha256: dict[str, str | None] = {}
    for index, artifact in enumerate(inputs):
        artifact = require_object(artifact, f"inputs[{index}]")
        role = require_string(artifact.get("role"), f"inputs[{index}].role")
        if role not in implementation.INPUT_ROLES:
            raise RequestError(f"{capability} does not accept input role {role}")
        if role in resolved:
            raise RequestError(f"duplicate input role: {role}")
        if artifact.get("cardinality") != "single":
            raise RequestError(f"input role {role} requires cardinality \"single\"")
        files = artifact.get("files")
        if type(files) is not list or len(files) != 1:
            raise RequestError(f"input role {role} requires exactly one file")
        file = require_object(files[0], f"inputs[{index}].files[0]")
        path = Path(require_string(file.get("path"), f"inputs[{index}].files[0].path"))
        if not path.is_file():
            raise RequestError(f"input file does not exist: {path}")
        if file.get("compression", "none") != "none":
            raise RequestError("the benchmark harness reads uncompressed inputs only")
        sha = file.get("sha256")
        if sha is not None and (type(sha) is not str or len(sha) != 64):
            raise RequestError(f"inputs[{index}].files[0].sha256 must be a 64-character hex string")
        resolved[role] = path
        declared_sha256[role] = sha
    missing = [role for role in implementation.INPUT_ROLES if role not in resolved]
    if missing:
        raise RequestError(f"{capability} requires inputs: {', '.join(missing)}")

    parameters = require_object(request.get("parameters", {}), "parameters")
    output_directory = Path(
        require_string(parameters.get("output_directory"), "parameters.output_directory")
    )
    if not output_directory.is_absolute():
        raise RequestError("parameters.output_directory must be an absolute path")
    for name in parameters:
        if name != "output_directory" and name not in implementation.PARAMETERS:
            raise RequestError(f"{capability} does not accept parameter {name}")
    if result_path.resolve().parent != output_directory.resolve():
        raise RequestError("the result path must live directly inside parameters.output_directory")
    if not output_directory.parent.is_dir():
        raise RequestError(f"output parent directory does not exist: {output_directory.parent}")
    for path in resolved.values():
        if path.resolve() == output_directory.resolve():
            raise RequestError("output directory must differ from every input")

    return {
        "job_id": job_id,
        "capability": capability,
        "implementation": implementation,
        "inputs": resolved,
        "declared_sha256": declared_sha256,
        "parameters": {k: v for k, v in parameters.items() if k != "output_directory"},
        "output_directory": output_directory,
        "result_path": result_path,
    }


def write_json_file(path: Path, document: dict[str, Any]) -> None:
    with path.open("w", encoding="utf-8", newline="\n") as handle:
        json.dump(document, handle, ensure_ascii=False, indent=2, sort_keys=True)
        handle.write("\n")
        handle.flush()
        os.fsync(handle.fileno())


def write_json_atomic(path: Path, document: dict[str, Any]) -> None:
    descriptor, temporary_name = tempfile.mkstemp(
        prefix=f".{path.name}.", suffix=".tmp", dir=path.parent
    )
    temporary_path = Path(temporary_name)
    try:
        with os.fdopen(descriptor, "w", encoding="utf-8", newline="\n") as handle:
            json.dump(document, handle, ensure_ascii=False, indent=2, sort_keys=True)
            handle.write("\n")
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(temporary_path, path)
    except BaseException:
        temporary_path.unlink(missing_ok=True)
        raise


def write_error_result_atomic(path: Path, document: dict[str, Any]) -> bool:
    if path.parent.is_dir():
        if path.exists():
            return False
        write_json_atomic(path, document)
        return True
    grandparent = path.parent.parent
    if not grandparent.is_dir() or path.parent.exists():
        return False
    staging = Path(tempfile.mkdtemp(prefix=".linxira-benchmark-error-", dir=grandparent))
    try:
        write_json_file(staging / path.name, document)
        if path.parent.exists():
            return False
        os.replace(staging, path.parent)
        return True
    finally:
        if staging.exists():
            shutil.rmtree(staging)


def software_provenance(implementation: Any) -> tuple[list[dict[str, str]], list[dict[str, Any]]]:
    entries: list[dict[str, str]] = [{"name": "CPython", "version": sys.version.split()[0]}]
    diagnostics: list[dict[str, Any]] = []
    for entry in implementation.software():
        entries.append(entry)
        if entry.get("package_id") == "biopython" and entry["version"] != LOCKED_BIOPYTHON:
            diagnostics.append(
                {
                    "code": "dependency_version_drift",
                    "severity": "warning",
                    "message": (
                        f"Biopython {entry['version']} is installed; the pack lock pins "
                        f"{LOCKED_BIOPYTHON}. Numbers remain comparable but the environment "
                        "disclosure records the installed version."
                    ),
                }
            )
    return entries, diagnostics


def run_benchmarked(config: dict[str, Any], started_at: str) -> dict[str, Any]:
    implementation = config["implementation"]
    capability: str = config["capability"]
    output_directory: Path = config["output_directory"]
    # The worker has already verified every declared hash against the file
    # content and re-checks the inputs after the pack exits, so the declared
    # digest is reused here: hashing a multi-terabyte input a third time would
    # only inflate the outer wall time of this backend. Standalone runs without
    # a declared digest compute it.
    input_sha256: dict[str, str] = {}
    for role, path in config["inputs"].items():
        declared = config["declared_sha256"][role]
        input_sha256[role] = declared.lower() if declared is not None else sha256_file(path)

    # Instrumentation brackets the analysis only. Interpreter start-up,
    # request validation and output writing are visible to the outer timer
    # (`/usr/bin/time -v` around the worker) but not to this one.
    started = time.perf_counter()
    result = implementation.run(config["inputs"], config["parameters"])
    wall_ms = (time.perf_counter() - started) * 1000.0
    peak = peak_rss_mb()

    software, diagnostics = software_provenance(implementation)
    diagnostics.append(
        {
            "code": SELF_REPORTED_CODE,
            "severity": "info",
            "message": json.dumps(
                {
                    "wall_ms": round(wall_ms, 3),
                    "peak_rss_mb": None if peak is None else round(peak, 3),
                    "instrument": "time.perf_counter + resource.getrusage(RUSAGE_SELF)",
                    "backend": BACKEND,
                },
                sort_keys=True,
            ),
        }
    )

    staging = Path(tempfile.mkdtemp(prefix=".linxira-benchmark-", dir=output_directory.parent))
    try:
        artifact_name = f"{capability}.result.json"
        staged_artifact = staging / artifact_name
        write_json_file(staged_artifact, result)
        lock_path = Path(__file__).resolve().parents[1] / "requirements.lock"
        envelope = {
            "schema_version": "2",
            "job_id": config["job_id"],
            "capability": capability,
            "status": "ok",
            "result": result,
            "artifacts": [
                {
                    "artifact_id": "benchmark-result",
                    "role": "benchmark-result",
                    "kind": "report",
                    "path": str(output_directory / artifact_name),
                    "format": "json",
                    "media_type": "application/json",
                    "size_bytes": staged_artifact.stat().st_size,
                    "sha256": sha256_file(staged_artifact),
                }
            ],
            "provenance": {
                "engine_version": PACK_VERSION,
                "execution_mode": "local-cpu",
                "core_version": core_version(),
                "started_at": started_at,
                "finished_at": utc_now(),
                "software": software,
                "input_sha256": input_sha256,
                "command": [
                    "python",
                    "src/benchmark_harness.py",
                    "--request",
                    "<request>",
                    "--result",
                    "<result>",
                ],
                "dependency_lock_sha256": sha256_file(lock_path),
            },
            "diagnostics": diagnostics,
        }
        write_json_file(staging / config["result_path"].name, envelope)
        if output_directory.exists():
            raise RequestError("output directory appeared while the analysis was running")
        os.replace(staging, output_directory)
        return envelope
    finally:
        if staging.exists():
            shutil.rmtree(staging)


def error_result(job_id: str, capability: str, message: str, started_at: str) -> dict[str, Any]:
    return {
        "schema_version": "2",
        "job_id": job_id,
        "capability": capability,
        "status": "error",
        "result": {},
        "artifacts": [],
        "provenance": {
            "engine_version": PACK_VERSION,
            "execution_mode": "local-cpu",
            "core_version": core_version(),
            "started_at": started_at,
            "finished_at": utc_now(),
        },
        "diagnostics": [{"code": "workflow_failed", "severity": "error", "message": message}],
    }


def parse_arguments(arguments: list[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Run the independent Python implementation of a Linxira Bio capability"
    )
    parser.add_argument("--request", required=True, type=Path, help="artifact-aware request JSON")
    parser.add_argument("--result", required=True, type=Path, help="result envelope JSON")
    return parser.parse_args(arguments)


def main(arguments: list[str] | None = None) -> int:
    options = parse_arguments(arguments)
    started_at = utc_now()
    job_id = "workflow-error"
    capability = "unknown"
    try:
        if not options.request.is_file():
            raise RequestError(f"request file does not exist: {options.request}")
        if options.request.resolve() == options.result.resolve():
            raise RequestError("result path must not alias the request file")
        with options.request.open("r", encoding="utf-8") as handle:
            document = json.load(handle)
        if type(document) is dict:
            if type(document.get("job_id")) is str and document["job_id"]:
                job_id = document["job_id"]
            if type(document.get("capability")) is str and document["capability"]:
                capability = document["capability"]
        config = validate_request(document, options.result)
        run_benchmarked(config, started_at)
        return 0
    except (OSError, json.JSONDecodeError, RequestError, RuntimeError, ValueError) as error:
        try:
            if options.request.resolve() != options.result.resolve():
                write_error_result_atomic(
                    options.result, error_result(job_id, capability, str(error), started_at)
                )
        except (OSError, RequestError):
            pass
        print(f"{PACK_ID}: {error}", file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())
